### DISCUSSION TITLE
argocd 2.5: --server-side-generate fails with helm charts ?

### DISCUSSION THREAD

**Discussion by dmpe** [argoproj/argo-cd#11315] [Q&A]
Hi all, 

I am using latest argocd 2.5.2 and attempting to use a new feature of `argocd app diff`, with a new `--server-side-generate` parameter. 
I have deployed both very simple kustomize apps, as well as (very complex - multiple deps) helm charts. And everything runs as expected. 

Now, I am following upgrade guideline - specifically https://argo-cd.readthedocs.io/en/stable/operator-manual/upgrading/2.4-2.5/#deprecated-client-side-manifest-diffs. 

My git repo has following structure:

```
tree -d -L 2
.
├── [drwxr-xr-x 4.0K Nov 16 13:09]  bin
├── [drwxr-xr-x  20K Nov 10 10:02]  lib
│   ├── [drwxr-xr-x 4.0K Nov 15 17:01]  helm -> here in each subfolder, we have apps like "keycloak" or "mattermost"
│   ├── [drwxrwxr-x 4.0K Nov 14 10:05]  kustomize -> same here: in each subfolder, we have apps like "argocd" or "jaeger-operator"
├── [drwxrwxr-x 4.0K Nov 15 17:01]  sbin
.... <skip unimportant stuff>
```

When creating an argocd app, we use following approach:

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: keycloak
  namespace: argocd
spec:
  destination:
    namespace: keycloak
    server: 'https://kubernetes.default.svc'
  source:
    path: lib/helm/keycloak
    repoURL: '<repo>'
    targetRevision: master
    helm:
      valueFiles:
        - cluster1_values.yaml
.....
```

The deployment works great, and `cluster1_values.yaml` being in the helm chart is used as "main entrypoint" for rendering the application (where we also store our internal configuration), i.e. it overrides other `values.yaml` file values. 

When doing app diff with an existing approach (the ones being deprecated), we current execute

```
argocd app diff keycloak --local "<path to git repo>/lib/helm/keycloak"  --local-repo-root <path to git repo>
```

This works. The diff is being displayed, with a warning about deprecation.

When now following documentation for `--sever-side-generate`, I use 

```
argocd app diff keycloak --local <path to git repo> --server-side-generate
```

and this fails with 

```
FATA[0002] rpc error: code = Unknown desc = `helm template . --name-template keycloak ...........................

failed exit status 1: Error: template: keycloak/charts/postgresql/templates/svc.yaml:10:20: executing "keycloak/charts/postgresql/templates/svc.yaml" at <{{template "common.names.fullname" .}}>: template "common.names.fullname" not defined
```

Interestingly, with a very simple kustomize file it works, but not with a single helm chart of ours (>=20). Even some more advanced kustomization apps also do not work. 

But at least with some other error:
```
DEBU[0003] Failed to apply normalization: error in remove for path: '/status': Unable to remove nonexistent key: status: missing value
DEBU[0003] Failed to apply normalization: error in remove for path: '/spec/preserveUnknownFields': Unable to remove nonexistent key: preserveUnknownFields: missing value
```

I have already activated debugging on repo-server and server components, but it shows very same errors as the client binary is getting. 

Anybody has tried this out and knows where the issue might lie?

Thanks. 

```
argocd: v2.5.2+148d8da
  BuildDate: 2022-11-07T17:06:04Z
  GitCommit: 148d8da7a996f6c9f4d102fdd8e688c2ff3fd8c7
  GitTreeState: clean
  GoVersion: go1.18.7
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.5.2+148d8da
  BuildDate: 2022-11-07T16:42:47Z
  GitCommit: 148d8da7a996f6c9f4d102fdd8e688c2ff3fd8c7
  GitTreeState: clean
  GoVersion: go1.18.8
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v4.5.7 2022-08-02T16:35:54Z
  Helm Version: v3.10.1+g9f88ccb
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.18.0
```

**Comment by dmpe** [comment]
As if there is lack of helm go templating .... i.e. `_helpers.tpl` files are not being processed. 
EDIT: I can confirm that when all templates are replaced with non-templated names, then it works nicely!

The normal helm buildin objects such as `{{ .Release.Name }}` or `{{ .Release.Service }}` work properly, but because tpl files are not pre-rendered, the whole chart diff fails.

@notfromstatefarm any thoughts? would you know why `tpl` seems to be skipped, when doing `argocd app diff --server-side-generate` ?

**Reply by dmpe** [reply]
Bump. Could anyone could take look, maybe pls?

**Reply by notfromstatefarm** [reply]
Will be looking into this shortly, sorry for the massive delay. I had to step back from Argo contributing for a couple of months.

**Reply by dmpe** [reply]
Hi @notfromstatefarm, thanks for getting back to us. My colleague found out that simply adding a `--local-include "*.*"` will fix the issue. Seems to us that `.tpl` files/suffix are not included in the filter and thus not uploaded to argocd.

```
--local-include stringArray   Used with --server-side-generate, specify patterns of filenames to send. Matching is based on filename and not path. (default [*.yaml,*.yml,*.json])
```
